server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -218,81 +214,56 @@ private synchronized void syncSegments(boolean isRetry) {
-                                    // along with marking minSeqNoToKeep, upload has succeeded completely.
-                                    shouldRetry = false;
-                                } catch (Exception e) {
-                                    // We don't want to fail refresh if upload of new segments fails. The missed segments will be re-tried
-                                    // in the next refresh. This should not affect durability of the indexed data after remote trans-log
-                                    // integration.
-                                    logger.warn("Exception in post new segment upload actions", e);
-                                } finally {
-                                    doComplete(shouldRetry);
-                                }
+                    // Create a map of file name to size and update the refresh segment tracker
+                    updateLocalSizeMapAndTracker(localSegmentsPostRefresh);
+
+                    // Create a map of file name to size and update the refresh segment tracker
+                    updateLocalSizeMapAndTracker(localSegmentsPostRefresh);
+                    CountDownLatch latch = new CountDownLatch(1);
+                    ActionListener<Void> segmentUploadsCompletedListener = new LatchedActionListener<>(new ActionListener<>() {
+                        @Override
+                        public void onResponse(Void unused) {
+                            boolean shouldRetry = true;
+                            try {
+                                // Start metadata file upload
+                                uploadMetadata(localSegmentsPostRefresh, segmentInfos);
+                                clearStaleFilesFromLocalSegmentChecksumMap(localSegmentsPostRefresh);
+                                onSuccessfulSegmentsSync(
+                                    refreshTimeMs,
+                                    refreshClockTimeMs,
+                                    refreshSeqNo,
+                                    lastRefreshedCheckpoint,
+                                    checkpoint
+                                );
+                                // At this point since we have uploaded new segments, segment infos and segment metadata file,
+                                // along with marking minSeqNoToKeep, upload has succeeded completely.
+                                shouldRetry = false;
+                            } catch (Exception e) {
+                                // We don't want to fail refresh if upload of new segments fails. The missed segments will be re-tried
+                                // in the next refresh. This should not affect durability of the indexed data after remote trans-log
+                                // integration.
+                                logger.warn("Exception in post new segment upload actions", e);
```

Comment:
nit on this comment - Is this retried on the next refresh or immediately?  This *will* impact the computation of staleness for replicas for segrep backpressure so we don't to wait until a future refresh.

Reply:
It is retried with exponential back-off strategy which starts with 1s interval.